Reference/Common-Pitfalls/index-v9.md

```diff
@@ -0,0 +1,426 @@
+---
+versionFrom: 9.0.0
+meta.Title: "Common Pitfalls and Anti-Patterns in Umbraco"
+meta.Description: "Information on common Pitfalls and Anti-Patterns in Umbraco"
+state: complete
+verified-against: rc-004
+update-links: true
+---
+
+# Common Pitfalls & Anti-Patterns
+
+_This section is ultra important! It will describe many common pitfalls that developers fall in to. Some of the anti-patterns mentioned here can bring your site to a grinding halt, cause memory leaks, or make your site unstable or perform poorly. Make sure you read this section - it might save your site! Another great source of information is Shannon Deminick's Codegarden 2021 talk about performance tuning which can be seen [here](https://www.youtube.com/watch?v=RM2EGXm6DYU&list=PLG_nqaT-rbpx6wIDr5ufUlbHkg6qB3sxH)_
+
+## Usage of Singletons and Statics
+
+Generally speaking if you are writing software these days you should be using Dependency Injection principles.
+If you do this, you probably aren't using Singletons or Statics (and for the most part you shouldn't be!). Since Umbraco 9 comes with depdenency injection out of the box, there really isn't any reason to use singletons or statics. It makes your code very difficult to test but more importantly using Singletons and Statics in your code make it very hard to manage, APIs become leaky and ultimately you'll end up with more problems than when you started.
+
+Depedency injection is available everywhere, and you can register your own services as well, additionally some resoureces are available through properties on certain base classes. For example, all Razor views that Umbraco creates expose an `UmbracoHelper` property you can access through `@Umbraco`, as well as a `SmidgeHelper` property. The other base classes that expose some things you might need like `UmbracoContext` are things like `SurfaceController`, but even here the services are initially gotten through DI, and you can inject further Umbraco and custom services that you might need.
+
+For more information about consuming and registering your own dependencies have a look at the [Dependency Injection](../../Reference/Using-Ioc/index-v9) documentation
```

Comment:
```suggestion
## Usage of Singletons and Statics

Generally speaking, if you are writing software these days you should be using Dependency Injection principles.
If you do this, you probably aren't using Singletons or Statics (and for the most part you shouldn't be!). Since Umbraco 9 comes with dependency injection out of the box, there really isn't any reason to use singletons or statics. It makes your code very difficult to test but more importantly using Singletons and Statics in your code makes it very hard to manage, APIs become leaky, and ultimately you'll end up with more problems than when you started.

Dependency injection is available everywhere, and you can register your own services as well, additionally, some resources are available through properties on certain base classes. For example, all Razor views that Umbraco creates expose an `UmbracoHelper` property you can access through `@Umbraco`, as well as a `SmidgeHelper` property. The other base classes that expose some things you might need like `UmbracoContext` are things like `SurfaceController`, but even here the services are initially gotten through DI, and you can inject further Umbraco and custom services that you might need.

For more information about consuming and registering your own dependencies have a look at the [Dependency Injection](../../Reference/Using-Ioc/index) documentation
```